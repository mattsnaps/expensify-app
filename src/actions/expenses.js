import database from "../firebase/firebase";
import { ref, push, get, remove } from "firebase/database";


// ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        const databaseRef = ref(database, '/expenses');
        return push(databaseRef, expense).then((ref) => {
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

// REMOVE_EXPENSE
export const removeExpense = ({ id: id } = {}) => ({
    id: id,
    type: 'REMOVE_EXPENSE'
});

export const startRemoveExpense = ({ id: id } = {}) => {
    return (dispatch) => {
        const databaseRef = ref(database, `expenses/${id}`);
        return remove(databaseRef).then(() => {
            dispatch(removeExpense({
                id
            }));
        });
    };
};

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

//SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        const expenseList = [];
        const databaseRef = ref(database, '/expenses');
        return get(databaseRef).then((snapshot) => {
            snapshot.forEach((childExpense) => {
                expenseList.push({
                    id: childExpense.key,
                    ...childExpense.val()
                });
            });
            dispatch(setExpenses(expenseList));
        });

    };
};
