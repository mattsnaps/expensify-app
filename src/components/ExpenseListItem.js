import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

/**
 *Alternative way to list out items.
 */
const ExpenseListItem_old = (props) => (
    <div>
        <h3>{props.description}</h3>
        <p>{props.amount} --- {props.createdAt}</p>
    </div>
);

//dispatch is in the props. same as props.dispatch
const ExpenseListItem = ({ description, amount, createdAt, id }) => (
    <div>
        <Link to={`/edit/${id}`}>
            <h3>{description}</h3>
        </Link>
        <p>
            {numeral(amount / 100).format('$0,0.00')}
            -
            {moment(createdAt).format('MMMM Do, YYYY')}
        </p>
    </div>
);

export default ExpenseListItem;
