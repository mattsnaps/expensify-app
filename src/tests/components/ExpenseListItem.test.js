import React from "react";
import { shallow } from "enzyme";
import ExpenseListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test('ExpenseListItem Success', () => {
    const testExpense = expenses[0];
    const wrapper = shallow(<ExpenseListItem {...testExpense}/>);
    expect(wrapper).toMatchSnapshot();
});