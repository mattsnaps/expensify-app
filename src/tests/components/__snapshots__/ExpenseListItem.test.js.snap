// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem Success 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Market
    </h3>
  </Link>
  <p>
    $1.95
    -
    December 31st, 1969
  </p>
</div>
`;
